daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala

```diff
@@ -77,8 +77,40 @@ final class Engine(nextRandomInt: () => Int) {
     * </ul>
     *
     *
-    * [[transactionSeed]] is the master hash used to derive node and contractId discriminator.
+    * [[submissionSeed]] is the master hash used to derive node and contractId discriminator.
     * If let undefined, no discriminator will be generated.
+    * The derivation scheme ensures the following properties:
+    * <ul>
+    * <li> For every subtree of the transaction, there is a seed such that all node and contract Id discriminators
+    *      can be derived from the seed based on the information in the subtree.  This allows to reinterpret
+    *      individual subtrees of a transaction.
+    * </li>
+    * <li> All node and contractId discriminators in the transaction are disjoint unless a hash collision occurs.
+    * </li>
+    * <li> Contracts with different stakeholders have different discriminators unless a hash collision occurs.
+    * </li>
+    * <li> Submissions with different submission time or participant Id lead to different node and contract Id discriminators
```

Comment:
From Canton's perspective, we don't need the dependence on submission time any more, neither on participant Id. So if no other DAML-on-X implementation needs this, we could remove it (and thus get rid of submission time).